PolyBFT consensus protocol

Stefan-Ethernal opened the pull request:
# Description

Placeholder PR for PolyBFT consensus protocol epic.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [ ] I have tested this code manually

### Manual tests

Please complete this section if you ran manual tests for this functionality, otherwise delete it

Nemanja0x committed:
Move PolyBFT to Edge (#774)

* Move PolyBFT to edge

Co-authored-by: Igor Crevar <crewce@gmail.com>
Co-authored-by: Sasa Prsic <sasa@ethernal.tech>
Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>
Co-authored-by: Victor Castell <victor@victorcastell.com>

igorcrevar committed:
Build block with Txs in PolyBFT (#789)


Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>
Co-authored-by: Sasa Prsic <sasa@ethernal.tech>

begmaroman committed:
EVM-48: Add RootChain helper (#773)

igorcrevar committed:
EVM-57 PolyBft state transactions introduction (#809)

* EVM-57 PolyBFT bridge enabled state txs

Co-authored-by: Sasa Prsic <sasa@ethernal.tech>

* EVM-57 PolyBFT bridge enabled state txs

Co-authored-by: Sasa Prsic <sasa@ethernal.tech>

* tidy.

* pr fixes

* pr fix 2

* small fix

* TransactionType should be populated in receipt

* nonce fix

* write change original tx fix

* createStateTransactionWithData change

* pr fix more

* revert ComputeHash for tx because id does not really matter

Co-authored-by: Sasa Prsic <sasa@ethernal.tech>

Commit:
Migrate polybft e2e framework to edge (#797)

Back-port the e2e framework + tests from v3 to edge as a separate set of tests.

Stefan-Ethernal committed:
v3 smart contracts integration (#796)

* Initial changes

* Init v3-contract submodule

* Uptime changes and linter fixes

* Use local build and test workflows (temporary)

* Test workflow expand with init git submodules

* Native transfer precompile

* CommitTransaction in blockBuilder

* adding workflow call secrets input (#799)

* Add BLS aggregated signatures verification precompile

* Change required gas for introduced precompiles

* Add console precompile

* Additional functions in bls package

* Register validator command

* Linter fixes, register subcommand for register validator command

* Adapt genesis command

* Cleanup

* Change addresses of sidechain contracts so that they don't clash with precompile addresses

* Marshal from address in transaction

* Remove CommitTransaction and use WriteTx instead

* Fix panic in register validator command

* Get rid of panics in register validator

* Increase gas limit for ChildValidatorSet initialization

* Cleanup GenesisPostHookFactory

* Fix register validator command

* Change gas limit to 100M for initialize child validator set

* Change make compile v3-contracts

* Recover CI (1st attempt)

* Recover CI (2nd attempt)

* Recover CI (3rd attempt)

* Cleanup Makefile (remove compilation of SCs when unneeded). Remove cloning submodules in lint workflow

* Recover lint

* Remove init submodules from test workflow

* Address some go security issues

* small make fix

* init root contracts command

* Remove rootchain contracts deployment from rootchain server command

* Provide new url for smart contracts

* Introduce constants for register validator and contracts root folder

* Comments

* Comments (part 2)

* Reference dev-ethernal commit from core-contracts

* Fix polybft e2e workflow

* Fix genesis cmd invocation from e2e framework

* contract_initializer fix

* Fix polybft e2e tests

* Comments (part 3)

* Populate validators balance based on provided genesis configuration.
Comments (part 4)

* Fix balance unmarshal for polybft.Validator

* Initialize native transfer contract

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>
Co-authored-by: Epikichi <78609649+epikichi@users.noreply.github.com>
Co-authored-by: Igor Crevar <crewce@gmail.com>

Stefan-Ethernal committed:
EVM-129 Implement bridge workflow (#828)

* Add bridge flag.

* Populate bridge config.

* Fix bridge address.

* rootchain panic fixes

* changes. deploying contracts to root still does not work in test

* fix deploying contracts and running bridge. Emit command change

* Read balance from genesis

* Remove Geth reference from Edge module

* bridge test fixes

* remove test-bootnode. add WithBootnodeCount to test-cluster. Change network_test to be more relaxed

Co-authored-by: Sasa Prsic <sasa@ethernal.tech>
Co-authored-by: Igor Crevar <crewce@gmail.com>

Stefan-Ethernal committed:
Transfer unit tests from v3 to Edge (#825)

* test-bridge fix + temporary fix

* Read balance from genesis

* state, state transaction, event tracker

* FSM tests (part 1)

* ConsensusRuntime tests

* FSM tests (part 2)

* FSM tests (part 3)

* SystemState tests

* ValidatorAccount and ValidatorSnapshot tests

* buildLogsFromReceipts test

* FSM tests (part 4)

* FSM fix remaining tests

* t.parallel

* Introduce parallel execution in tests (part 2)

* Lint fixes

* Rename test helpers file to exclude it from build

* Add test suffix to mocks

* Install solidity compiler on test workflow

* Assign execute permission to setup-ci.sh script

* polybft tests

* all tests merged

* setting epoch snapshot added

* Lint fixes

* Use constant for IstanbulExtraVanity

* Description of polybft test

Co-authored-by: Igor Crevar <crewce@gmail.com>
Co-authored-by: Nemanja0x <nemanja@ethernal.tech>
Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

goran-ethernal committed:
[Checkpointing] `ExitEvent` storage and proof generation (#838)

* Exit event storage and query functions

* ExitProof generation

* Reorganize tests

* Optimize key composition

* Extract common function for decoding state sync and exit events

* Comments fix

* Comments fix

* [Checkpointing] JSON RPC point for generating exit proofs (#844)

* JSON RPC point for generating exit proofs

* Comments fix

Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>

Stefan-Ethernal committed:
Checkpoint data signing (#840)

* Add CheckpointData to Extra

* Don't store chainID, blockNumber and blockHash to the blockchain

* Set checkpoint data to the Header.Extra

* Fix and simplify FSM unit tests

* getSignHash static function, fix verify header implementation

* Increase block waiting time in bridge test

* Remove unnecessary condition

* Set checkpoint to each extra

* Sign checkpoint hash instead of aggregated hash

* Make nextValidatorsHash same as currentValidatorsHash for non-epoch ending blocks

* Set event root to the checkpoint object

goran-ethernal committed:
[Checkpoints] Parse receipt logs and insert new exit events (#852)

* Parse receipt logs and insert exit events

* Comments fix

* Comments fix

* Comments fix

* Remove TODO

* Comments fix

* Comments fix

* Comments fix

Stefan-Ethernal committed:
Consolidate BLS verification ABI type (#867)

Stefan-Ethernal committed:
Checkpoint submission (#861)

* Change error messages for checkpoint hash calculation

* Determine checkpoint blocks

* Checkpoint submission (draft)

* Introduce checkpoint manager

* Lint fixes

* Fix tests

* Checkpoint manager checkpoint data ABI encode unit tests

* Unit test for getCurrentCheckpointID

* Unit test for submitCheckpoint

* Remove leftover code

* Comments

* Determine epoch ending blocks by comparing epoch numbers in adjacent headers

* Micro optimization

* Skip querying first block for end of epoch

* Optimization

* Simplify checkpoints offset

* Adapt comment

* Checkpoint block detection optimization (#872)

* Simplify checkpoint block detection

* Resolve conflicts and pass tests

Co-authored-by: Boris <boris@Boriss-MacBook-Pro.local>

* Simplify

* Submit pending checkpoints in unit test

* Extract parentHeader retrieval outside of the for

* Submit end of epoch blocks for pending checkpoints

* Update submitCheckpoint comment

Co-authored-by: Victor Castell <victor@polygon.technology>

* Feedback

* Separate import declarations

* Create var block for `parentExtra` and `parentHeader`

Co-authored-by: Victor Castell <victor@polygon.technology>

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>
Co-authored-by: Boris <boris@Boriss-MacBook-Pro.local>
Co-authored-by: Victor Castell <victor@polygon.technology>

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Mock Host implementation in native transfer test (#887)

* Mock Host implementation in native transfer test

* Lint fix

* Fix polybft e2e tests

Commit:
Implement the logic for quorum check (#837)

* Implement the logic for quorum check

Depending on the message type it will implement different kinds of quorum checks.

* Remove MaximumFaulty

Stefan-Ethernal committed:
Integrate go-ibft consensus engine to polybft consensus protocol (#891)

* EVM 149 - syncer go ibft coordination (#843)

* consensus runtime backend metohod (not buildable)

* transport

* polybft - run cycle

* transport fix 2

* small changes

* less complex ibft consennsus wrapper.Example: https://replit.com/@crewce/Newest#main.go

* small changes

* runSequence better

* remove pbftTransportWrapper

* noone likes temp variables

* Comment fixed

* jm fixes

Co-authored-by: Igor Crevar <crewce@gmail.com>

* EVM-151 Use (or not) edge hooks support (#849)

* FSM that supports go-ibft backend interface (#865)

* consensus runtime backend metohod (not buildable)

* transport

* polybft - run cycle

* transport fix 2

* less complex ibft consennsus wrapper.Example: https://replit.com/@crewce/Newest#main.go

* Implement MessageConstructor interface.

* Tidy linter errors.

* runSequence better

* remove pbftTransportWrapper

* Add quorum size implementation.

* Remove one part of pbft.NodeID reference.

* Add some comments.

* IsValid, Build proposal on FSM

* Add block validation.

* Add stub for IsValidSender method.

* Sender validation is always true.

* fsm - Insert block

* simplify build proposal

* Tidy logs.

* IsValidSender

* pbft.Proposal removed

* Tidy errors.

* restartEpoch if epoch is nil, remove hook in fsm

* initialization fix

* Fix compare.

* quorum fixes and error propagation

* evm-151 changes + additional needed in this task

* Increase test timeout.

* bring back some part of fsm

* just execute syn test 100 times in a row

* Tidy part of errors.

* Revert e2e test.

* Remove unused code and move message creator interface implementation.

* minor fix

* Add comments.

* code organization - valdateSender + recoverAddress

* Fix compile after merge.

* small change

* restart epoch on better place. lastBuiltBlock and epoch should not be nil

* small reorg in polybft

* pr fix

* Update comments

* pr fixes no 2

* Add voting power to the validators (#880)

* Add voting power to the validator account

* Rename ValidatorAccount to ValidatorMetadata

* Set voting power to the existing test

* Comment fixes

* Rename files

* Address comments

* Remove extra space from the comment

* pr fixes

* Rename GetValidatorAccount to GetValidatorMetadata

Co-authored-by: Nemanja0x <nemanja@ethernal.tech>
Co-authored-by: Igor Crevar <crewce@gmail.com>
Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>
Co-authored-by: Stefan Negovanović <93934272+Stefan-Ethernal@users.noreply.github.com>

* Accommodate go-ibft `BuildProposal` updated function (#893)

* Populate round info and change proposal hash

* Adopt currentRound from BuildProposal

* Fix logs

* Fix proposal hash calculation inside IsValidProposalHash

* Update go-ibft version

* Igor's feedback addressed

* Merge fix

* EVM-187 Handling bundles on both sync and consensus (#882)

* EVM-187 Handling bundles on both sync and consensus

* consensus_runtime tests

* UT fix

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

* Implement HasQuorum function in consensusRuntime

Co-authored-by: Nemanja Nedic <93836814+Nemanja0x@users.noreply.github.com>
Co-authored-by: Igor Crevar <crewce@gmail.com>
Co-authored-by: Saša Pršić <93726535+0xSasaPrsic@users.noreply.github.com>
Co-authored-by: Nemanja0x <nemanja@ethernal.tech>
Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Remove Polybft dummy smart contracts (#909)

Stefan-Ethernal committed:
Change ValidatorMetadata methods to pointer receivers (#911)

igorcrevar committed:
EVM-152 Fix Unit Tests (#910)

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

epikichi committed:
adding devnet-v3 workflow for parity testing. Adding needed secrets for other workflows (#916)

Commit:
Introduce local development cluster script (#912)

* Introduce local development cluster script

Stefan-Ethernal committed:
Fix assigned boot nodes port number in genesis (#930)

Commit:
Local cluster - adapt port for ibft to polybft (#931)

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Fix unit test for event tracker (#935)

* Fix linters

* Fix TestEventTracker_TrackSyncEvents

* Comment

igorcrevar committed:
EVM-232 Keep polybft bls and ecdsa keys in separate secrets manager namespaces (#953)

Stefan-Ethernal committed:
Track updates for validators (#939)

* Implement equality checks for PublicKey and ValidatorMetadata.
Expand ValidatorSetDelta

* Handle updates in ApplyDelta

* Additional tests

epikichi committed:
Epikichi/edge 973 change v3 parity workflow to devnet (#957)

* reflecting changes to use ephemeral environment instead

* Epikichi/edge 964 implement multiple erc mode support (#954)

* adding support for ERC20 and ERC721 mode for Pandora's Box stress testing

* adding in default for devnet ephemeral anticipation

* fixing markdown format

* reverting for a separate PR

* adding changes for devnet ephemeral workflow to allow dispatching

* adding correct job names, removed dependencies since concurrency protection is set on Pandora' workflow env-wide

* reflecting chnages from develop

igorcrevar committed:
EVM-220 TestClusterBlockSync/BLS fails in voting power branch (#926)

0xSasaPrsic committed:
Fix data race to polybft unit test (#965)

* Fix some race conditions.

Stefan-Ethernal committed:
Polybft voting power (#944)

voting power feature

Stefan-Ethernal committed:
Align `v3-parity` with `develop` branch (#971)

* Epikichi/edge 964 implement multiple erc mode support (#954)

* adding support for ERC20 and ERC721 mode for Pandora's Box stress testing

* adding in default for devnet ephemeral anticipation

* fixing markdown format

* reverting for a separate PR

* introducing workflow dispatch trigger for Devnet V3 workflow. (#927)

* introducing workflow dispatch trigger for Devnet V3 workflow.

* adding ephemeral environment naming

* adding pandora modes

* Add performance tests to devnet deployments (#918)

* added performance tests to devnet deployment
* removing depricated loadbot performance tests

* Docker setup: Wait for genesis file to exist before starting nodes (#949)

* Docker setup: Wait for genesis file to exist before starting nodes

* Docker setup: don't run init if genesis file already exists

* adding conditional dependency for compose to wait for successful  node

Co-authored-by: epikichi <ji@epik.dev>

Co-authored-by: Epikichi <78609649+epikichi@users.noreply.github.com>
Co-authored-by: ZeljkoBenovic <47507986+ZeljkoBenovic@users.noreply.github.com>
Co-authored-by: mediremi-antithesis <107197847+mediremi-antithesis@users.noreply.github.com>
Co-authored-by: epikichi <ji@epik.dev>

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Fix e2e tests

Stefan-Ethernal committed:
Set address to the result instance on applyCreate

Commit:
Remove State Sync execution step (#964)

* Remove commitment execution related stuff
* Add the RPC endpoint to obtain StateSync Proofs
* Rootchain data to gitignore
* Return the state sync along with proof
* Better params for cluster script
* E2E test executing state syncs
* Fix tests that rely on bundle execution
* Remove commitment execution related stuff

igorcrevar committed:
Replace BLS library with github.com/kilic/bn254 (#978)

goran-ethernal committed:
Integrate with CheckpointManager.sol (#952)

* Integrate with CheckpointManager.sol

* Hex encode ECDSA private key

* Skip failing tests for checkpoint manager

* Fix submit function signature

* Remove logs and set From field to submit checkpoint transaction

* Pass signer key to rootchain.SendTxn function

* Pull console precompile Geth docker image

* Populate voting power on checkpoint manager initialization

* Hash change

* Minor fixes (ordering of abi ancoded parameters)

* Populate voting power on genesis block

* Fix balance scaling

* Fix TestCheckpointManager_isCheckpointBlock

* Remove sorting by addresses on validators snapshot calculation

* Sort validators by addresses on genesis header population

* Sort validators by addresses on rootchain initialization

* Logging...

* Update core-contracts

* Fix checkpoint manager unit tests

* Change comment + unit tests for helper functions

* Revert usage of Geth console fork

* Include Geth console fork again

* Rename latestCheckpointBlockNumber to currentCheckpointBlockNumber

* Remove logs and revert to use official Geth image for rootchain server

* Code cleanup

* Add meaningful logs to the checkpoint manager

* not buildable

* full functional bn254

* Use proper domain when initializing CheckpointManager contract

* Fix go mod

* Revert helper bash scripts

* Fix e2e tests

* Address Igor's comment

* Provide `epochSize` to the `ChildValidatorSet` (#983)

* Provide epochSize to the ChildValidatorSet

* Rename newBls map key

* Rename parameters

* Point to our branch in core-contracts

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>
Co-authored-by: Nemanja0x <nemanja@ethernal.tech>
Co-authored-by: Igor Crevar <crewce@gmail.com>
Co-authored-by: Stefan Negovanović <93934272+Stefan-Ethernal@users.noreply.github.com>

goran-ethernal committed:
Merge remote-tracking branch 'origin/develop' into feature/v3-parity

Stefan-Ethernal committed:
Change genesis premine logic (#981)

0xSasaPrsic committed:
Fix data race in fsm (#982)

* Fix data race in several places accessing fsm and fix copy validators function.

* tidy linter.

* Fix tests after changes and user RLock.

Commit:
Removing `GenesisTarget` and adding the `NodeID` field to `Validator` struct (#985)

* Refactored GenesisTarget removing it and adding the NodeID field to Validator struct.
* Refactored generatePolyBFTConfig to use the new struct and simplify the code and take into account validators passed from params
* Add NodeID as a field needed for validators passed from params
* Tune timeout for bridge test

Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>

Stefan-Ethernal committed:
Provide rootchain IP address from CLI (#989)

* Refactor RootchainInteractor abstraction

* Lint fix

* Provide JSON RPC address flags

* Remove lock from rootchain interactor.
Send transaction remove nonce parameter.

* Address comments

* Reduce amount of affected files

* Implement TxRelayer

* Address comments (part 2)

* Create txn with nonce set

* Option pattern in TxRelayer

* Use txRelayer in e2e tests

* Let TxRelayer take care of setting the nonce

* Fix tx pool e2e tests

begmaroman committed:
Added london hardfork to the chain config (#999)

Stefan-Ethernal committed:
Fix pending checkpoints submission (#992)

* Fix epoch ending blocks detection in pending checkpoints

* Update TestCheckpointManager_submitCheckpoint

* Remove redundant signerAddress

* Make unit test more understandable

Stefan-Ethernal committed:
Update core-contracts (#1007)

goran-ethernal committed:
Bundle build fix (#1014)

goran-ethernal committed:
Check for nil proposal (#1009)

ferranbt committed:
Remove gasLimit and gas fields from txrelayer call (#1020)

ferranbt committed:
Add default gas and gasPrice to local transactions (#1021)

ferranbt committed:
Use geth console fork as default with flag to disable (#1022)

Stefan-Ethernal committed:
Checkpoint submission e2e test (#990)

* Checkpoint submission e2e test

* Utilize TxRelayer in test

* Fix init contracts command

* Fix e2e test

Commit:
Merge branch 'develop' into feature/v3-parity

JekaMas committed:
git modules: shallow and fix branch (#987)

* shallow and fix branch

* feat-polybft-release

Stefan-Ethernal committed:
Remove bool return value from deliverMessage signature (#1027)

0xSasaPrsic committed:
Fix/immutable trie race (#1017)

* Add lock to trie state.

Stefan-Ethernal committed:
Use total stake for voting power (#1038)

* Use total stake as voting power

* Lower log level to debug log in local cluster script

* Fix unit test

Stefan-Ethernal committed:
RLP Marshal/Unmarshal update field in `ValidatorSetDelta` (#1026)

* RLP Marshal/Unmarshal update field in ValidatorSetDelta

* Comment + additional unit test

Stefan-Ethernal committed:
Remove unused fields (#1039)

begmaroman committed:
Introduce metrics for BoltDB for the state (#1045)

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Expose `PolyBFT` metrics to Prometheus API (#1040)

* Introduce metrics to polybft

* Add comment

* Update comment

* Remove total state syncs metrics

* Include metrics prefix and rename file

* Address comment

* Fix consensus runtime tests

goran-ethernal committed:
EVM-308 Add commands we are missing (#1050)

* Added new commands

* Merge params and result files

* Reorganize commands

* Comments fix

Stefan-Ethernal committed:
Fix loop closure violations in unit tests (#1053)

* Fix loop closures

* Make TestExtra_CreateValidatorSetDelta_Cases deterministic

* Avoid for loop when assigning voting power

* Minimize changes

* Fix

b00ris committed:
Integrate L2 exit workflow (#1024)

* e2e exit test

* fix lint

* add static L1 exit contract data

* add contract comilation to e2e tests

* fix ci

* save artifacts in client

* fix tests

* review fix

* review fix

* Update consensus/polybft/contractsapi/gen/main.go

Co-authored-by: Stefan Negovanović <93934272+Stefan-Ethernal@users.noreply.github.com>

* review fix

Co-authored-by: Stefan Negovanović <93934272+Stefan-Ethernal@users.noreply.github.com>

Stefan-Ethernal committed:
Add pending checkpoints assertions to checkpoint submission e2e test (#1052)

Stefan-Ethernal committed:
Decouple rootchain contracts deployment from development environment (#1005)

* Rootchain init contracts read validators from the genesis configuration.
Introduce manifest.json.

* Introduce admin-key flag to the rootchain init-contracts

* Revert polybft params changes

* Fix comment

* Populate rootchain admin address to manifest

* Update Manifest struct and adapt init-contracts workflow

* Remove sidechain smart contract addresses from genesis spec

* Update genesis command and cleanup genesis from unnecessary parameters

* Remove rootchain contracts hardcoded addresses

* Decouple emit command from contant StateSender address

* Manifest command

* Polybft manifest command

* Adapt e2e tests workflow

* Update comment

* Adapt cluster script and change flags descriptions in manifest command

* Comments

* Adapt TestE2E_CheckpointSubmission

* Remove manifest.json in cluster script

* Initial version of polybft readme

* Update readme with a precondition step

* Address comments (part 1)

* Update manifest command short help

* Comments

* Use sidechain addresses from the polybft config in fsm UTs

* Minimize changes

* Merge fix

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

igorcrevar committed:
EVM-302  Save proposer priorities after each block (#1018)


* EVM-302 Save proposer priorities after each block

Co-authored-by: Nemanja0x <nemanja@ethernal.tech>
Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>
Co-authored-by: Sasa Prsic <sasa@ethernal.tech>

Stefan-Ethernal committed:
Remove GossipMessage (#1057)

Stefan-Ethernal committed:
Consolidate `ExtendByteSlice` implementations (#1058)

* Consolidate ExtendByteSlice implementations

* Add comment and fix

* Introduce trim parameter (fix unit tests)

* Update bitmap unit tests

* Address comment

* Add a unit test

Stefan-Ethernal committed:
Remove readContractBytecode (#1061)

Nemanja0x committed:
Evm 303: Fix race conditions after a block is inserted (#1047)


Co-authored-by: Igor Crevar <crewce@gmail.com>
Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>
Co-authored-by: Sasa Prsic <sasa@ethernal.tech>

0xSasaPrsic committed:
Feature/validator set changes (#1048)

* Update priorities when validator set changes.

Commit:
Remove code related to bundles for state syncs (#1013)

Related to RFC-139, we're removing bundles concept from the codebase.
We're introducing the latest Smart Contracts version in this PR

ferranbt committed:
Create a full block object during polybft (#1063)

* Create a full block object during polybft

* Remove print

* Remove now

* Fix tests

Stefan-Ethernal committed:
Fix segmentation fault panic on `ValidateSender` (#1054)

* Add nil check for FSM

* Different approach: consensus runtime validates message sender.
Remove Includes function from validatorSet.

* Subscribe IBFT topic only after epoch is initialized

* Renames and comment fixes

* Revert sender validation to the fsm

* Rebase fix

* Add log

goran-ethernal committed:
RFC173 Checkpoint consensus issue (#1064)

* RFC 173 changes

* Comments fix

* Comments fix

Stefan-Ethernal committed:
Fix uptime total blocks calculation (#1069)

* Fix uptime total blocks calculation

* Update test for uptime calculation

* Simplify

* Simplify unit test

* Typo fix

Commit:
Implement the relayer (#1044)

* Implement the relayer
* The statesyncrelayer package will reside inside polybft

0xSasaPrsic committed:
Immutable trie simplification (#1068)

* Refactor immutable trie.

* Remove state from trie.

* Tidy.

* Remove storage from trie.

Stefan-Ethernal committed:
Remove voting power scaling (#1060)

* Change voting power from uint64 to big.Int

* Provide desired scaling factor to validator set

* Fix unit tests after rebase

* Fix validatorSetToABISlice comment

* Introduce relative voting power

* Change ProposerPriority to big.Int

* Fix log

* Fix computeMaxMinPriorityDiff

* Remove max total voting power hard limit

* Use truncated division for rescaling priorities

* Add ceil to quorum size calculation

* Revert unnecessary changes

* Remove safe_math

* Compress quorum calculation

Co-authored-by: Victor Castell <victor@polygon.technology>

* Extract helper function for big int with ceil division

Co-authored-by: Victor Castell <victor@polygon.technology>

epikichi committed:
Epikichi/edge 988 add polybft consensus support to edge (#1051)

* Added support for polybft consensus docker cluster
* added customisation options
* added/edited readme

Co-authored-by: ZeljkoBenovic <sakal87@gmail.com>
Co-authored-by: ZeljkoBenovic <47507986+ZeljkoBenovic@users.noreply.github.com>

Stefan-Ethernal committed:
Implement `TestE2E_Bridge_ChangeVotingPower` (#1075)

* Implement TestE2E_Bridge_ChangeVotingPower

* Increase epoch rewards

* Add log to the fsm.getCurrentValidators

* Move default epoch reward to genesis cmd

* Dump validators to log only when debug logs enabled

* Multiple validators are staking

Stefan-Ethernal committed:
Remove unnecessary panics from polybft (#1077)

* Register validator remove panics and expose json rpc flag

* Remove panic from blockchainWrapper

* Remove json rpc flag from register validator cmd

* Polybft log instead of panic if block synchronization fails

* Remove unused vars

* Handle createValidatorSetDelta error

* Rename test

* Revert panic in stateProvider

goran-ethernal committed:
Dynamic epoch size (#897)

* Decouple uptime calculation from fixed epoch size

* Tests fix

* Rename isEndOfEpoch

* Return error on apply delta to validator set

* Comments fix

* Comments fix

* Comments fix

* Comments fix

* Comments fix

* Bug fix

* Comments fix

* Rebase fix

* Rebase fix

* Rebase fix

* Rebase fix

* Modify snapshot calculation for dynamic epoch (#997)

* Modify snapshot calculation for dynamic epoch

* Comments fix

* Comments fix

* Move tests to new file

* Comments fix

* Rebase fix

* Comments fix

* Comments fix

ferranbt committed:
Feature - State Sync manager (#1067)

* initial work

* Rework on event tracker and tests

* More stuff

* Migrate last test

* More changes

* Remove unused test

* Use correct core-contracts version

* Remove logs

* Use eventTracker from tracker package, code cleanup, bug fixes

* Code cleanup

* Code cleanup

* Comments fix

* Dummy state sync manager when bridge is not enabled

* Comments fix

* Comments fix

* Rebase fix

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

Commit:
Core contracts to dev-audit (#1095)

* Core contracts to dev-audit
* Fix event name

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/0xPolygon/polygon-edge?pullRequest=899) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/0xPolygon/polygon-edge?pullRequest=899) before we can accept your contribution.<br/>**0** out of **11** committers have signed the CLA.<br/><br/>:x: Nemanja0x<br/>:x: igorcrevar<br/>:x: begmaroman<br/>:x: vcastellm<br/>:x: ferranbt<br/>:x: JekaMas<br/>:x: epikichi<br/>:x: 0xSasaPrsic<br/>:x: Stefan-Ethernal<br/>:x: goran-ethernal<br/>:x: b00ris<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/0xPolygon/polygon-edge?pullRequest=899) it.</sub>

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Implement `Stringer` interface in `AccountSet` (#1103)

* Implement stringer in AccountSet

* Use strings.Join instead of defining buffer explicitly

Stefan-Ethernal committed:
Add disclaimer message when running polybft consensus protocol (#1107)

* Add disclaimer message when running polybft consensus protocol

* Print disclaimer message to logger and CLI outputter

Stefan-Ethernal committed:
Register validator e2e test (#1084)

* Initial approach

* Rely on system state in TestE2E_Consensus_RegisterValidator

* Add stake and balance flags to register validator cmd

* Remove debug log

* Add appropriate assertions

* Revert scientific notation changes

* Add comments

* Remove grpc flag from register validator cmd

* Delegation and undelegation e2e test (#1086)

* Initial approach

* Delegation logic

* Minor changes to unstake and withdraw commands

* Withdrawal and undelegate in test

* Change unstake cmd description

* Query for current block number before outputting delegator info

Stefan-Ethernal committed:
Handled go routines in event tracker (#1080)

* Manage goroutines in EventTracker

* Stop StateSyncRelayer

* Remove closeCh from polybft and use context instead

* Create named logger in the event tracker itself

* Update consensus/polybft/polybft_test.go assertion message

Co-authored-by: Victor Castell <victor@polygon.technology>

* Remove context field from consensusRuntime

* Isolate changes (revert changes in polybft to rely on channel)

* Add function comments

* Fix test

Co-authored-by: Victor Castell <victor@polygon.technology>

goran-ethernal committed:
Multiple state sync commitments per epoch (#1083)

* Initial changes

* e2e test

* Code cleanup

* Comments fix

* Lock fix

* Comments fix

* Comments fix

* Comments fix

* Comments fix

* Update comment in state_sync_manager

Co-authored-by: Stefan Negovanović <93934272+Stefan-Ethernal@users.noreply.github.com>

Commit:
Merge branch 'develop' into feature/v3-parity

goran-ethernal committed:
Return FullBlock from syncer (#1110)

* Return FullBlock from syncer

* Comments fix

* Comments fix

* Extract common parts of WriteBlock and WriteFullBlock

* Revert "Extract common parts of WriteBlock and WriteFullBlock"

This reverts commit 62166e6877f5a6a6f4a780d7047511caa530b8df.

Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>

goran-ethernal committed:
EVM-356 Update checkpoint out of fsm (#1113)

* Move logic from fsm to checkpoint manager

* PostBlock and PostEpoch in checkpointManager

* Function comment fix

* Call PostEpoch on checkpointManager

* Test fix

* e2e tests fix

* Comments fix

* Comments fix

* Change test

* Comments fix

goran-ethernal committed:
Move BuildEventRoot to checkpointManager (#1115)

Stefan-Ethernal committed:
Initialize `closeCh` in `stateSyncManager` (#1116)

* Initialize closeCh in stateSyncManager

* Rename test

* Add test for state sync relayer

goran-ethernal committed:
Move GenerateExitProof to checkpointManager (#1118)

Stefan-Ethernal committed:
Invert logical condition in `WaitUntil` functions (#1123)

* Invert logic for bridge WaitUntil function

* Invert WaitUntil logic in TestCluster

* Change condition in TestE2E_Bridge_ChangeVotingPower

Stefan-Ethernal committed:
Fix flaky register validator e2e test (#1122)

* Invert logic for bridge WaitUntil function

* Invert WaitUntil logic in TestCluster

* Change condition in TestE2E_Bridge_ChangeVotingPower

* Fix flaky test

* Query validators multiple times

Nemanja0x committed:
TestE2E_Bridge_L2toL1Exit to work with multiple users  (#1132)

TestE2E_Bridge_L2toL1Exit to work with multiple users

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

goran-ethernal committed:
Use PostBlock handler in ProposerCalculator (#1139)

Stefan-Ethernal committed:
Add missing cluster stop invocations in e2e tests (#1138)

vcastellm commented:
Let's mark this Ready for review

ZeljkoBenovic committed:
Optimized docker build in docker-compose (#1135)

vcastellm commented on contracts/system_addresses.go:
```diff
@@ -0,0 +1,27 @@
+	// BLSAggSigsVerificationPrecompile is an address of BLS aggregated signatures verificatin precompile
+	BLSAggSigsVerificationPrecompile = types.StringToAddress("0x2030")
+	// ConsolePrecompile is and address of Hardhat console precompile
+	ConsolePrecompile = types.StringToAddress("0x000000000000000000636F6e736F6c652e6c6f67")
```
Do we want to leave this?

Stefan-Ethernal replied:
I think it is convenient to have console precompile enabled (though it is coupled to the development environment strictly). We may need a broader understanding of how we will draw lines between production vs staging vs development environment in Polygon Edge in the future.

vcastellm commented on Makefile:
```diff
@@ -11,10 +11,11 @@ bindata:
 protoc:
 	protoc --go_out=. --go-grpc_out=. ./server/proto/*.proto
-	protoc --go_out=. --go-grpc_out=. ./protocol/proto/*.proto
+	#protoc --go_out=. --go-grpc_out=. ./protocol/proto/*.proto
```
This comment

vcastellm commented on blockchain/blockchain.go:
```diff
@@ -679,44 +679,48 @@ func (b *Blockchain) WriteHeadersWithBodies(headers []*types.Header) error {
 // VerifyFinalizedBlock verifies that the block is valid by performing a series of checks.
 // It is assumed that the block status is sealed (committed)
-func (b *Blockchain) VerifyFinalizedBlock(block *types.Block) error {
+func (b *Blockchain) VerifyFinalizedBlock(block *types.Block) (*types.FullBlock, error) {
```
Something to keep an 👁️

vcastellm commented on state/immutable-trie/snapshot.go:
```diff
@@ -72,7 +74,76 @@ func (s *Snapshot) GetCode(hash types.Hash) ([]byte, bool) {
 }
 
 func (s *Snapshot) Commit(objs []*state.Object) (state.Snapshot, []byte) {
```
Watch out this

vcastellm commented on types/rlp_marshal.go:
```diff
@@ -186,5 +190,13 @@ func (t *Transaction) MarshalRLPWith(arena *fastrlp.Arena) *fastrlp.Value {
+		vv.Set(arena.NewBytes([]byte{byte(t.Type)}))
+	}
+
+	if t.IsStateTx() {
```
In theory this shouldn't affect `ibft` as there are not state txns

vcastellm reviewed (commented):
Overall I just see some possible interactions with old code to keep an eye but nothing major.

Commit:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Validator unstake e2e test (#1112)

* Initial approach

* Tweak test

* Invert WaitUntil condition

* Premine balance to child smart contracts

* Working withdrawal

* Remove redundant inline function

* Pre-allocate ChildValidatorSet in genesis with initial validator set total stake

* Rebase fix

* Address comments

Commit:
E2E Relayer (#1131)

E2E Tests for Relayer

We need to check if StateSyncs are being executed correctly by one of the nodes of the cluster, so we are removing the "manual" execution previously present in e2e test and relay on the Relayer feature to do it as normal operation.

goran-ethernal committed:
BuildProofs for missed commitments (#1141)

* BuildProofs for missed commitments

* Comments fix

* Comments fix

igorcrevar committed:
EVM-360 Invalid state root error on v0.7.1-alpha2 (#1133)

Change block builder Reset to include timestamp for current block and call Reset each time before proposal is created and not just once when we are creating fsm

Stefan-Ethernal committed:
Rely on contracts api package when dealing with ABI abstractions (#1144)

* Add remaining smart contract artifacts

* Use ABI functions, events and types from Artifact abstractions

* Unexport stateSyncABIType variable

* Extract StateSynced event from correct contract

* Rename parameters so that they match smart contract spec

* Fix unstake and undelegate

* Typo fix

* Fix undelegate e2e test

begmaroman committed:
Align RLP marshaling logic according to eip-2718 (#1114)

igorcrevar committed:
EVM-360 e2e test (#1158)

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

igorcrevar committed:
EVM-353 Different voting power property based test (#1149)

Stefan-Ethernal committed:
Remove GetNonce function (#1165)

Stefan-Ethernal committed:
Expose ValidatorInfo struct (#1162)

ferranbt committed:
Feature contractsapi gen (#1062)

* Add initial gen api

* Remove method suffix

* Add contracts to the main function

* Add struct encoding

* Small fix

* Generate contract namespace (placeholder)

* Handle leading underscore in field name

* Comments fix

* Comments fix

* Take events from artifact

* Event name fix

* Remove contracts stubs from generated code

* Remove unnecessary code

* Real type names

* Comments fix

* Comments fix

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

Nemanja0x committed:
TestCommitEpoch unit test (#1167)

TestCommitEpoch

igorcrevar committed:
EVM-395 Implement RFC-175 BLS integration (#1161)

* EVM-395 Implement RFC-175
* Add hash to point unit tests
* Avoid random secret keys with zero value
* Check point at infinity and subgroup

Co-authored-by: Ferran Borreguero <ferranbt@protonmail.com>

begmaroman committed:
Removed legacy logic from tx pool (#1171)

Stefan-Ethernal committed:
Introduce test smart contracts as artifacts (#1163)

* Introduce L1DummyStateReceiver as artifact

* Extract TestL1StateReceiver smart contract to separate file and embed it

* Include TestWriteBlockMetadata artifact

* Introduce constant for test-contracts dir

Stefan-Ethernal committed:
Checkpoint data validations (#1160)

* Validate checkpoint data

* Separate Validate and ValidateBasic checkpoint data validations

* Summarize logical conditions

* Vefify state transactions before writing commit epoch transaction

* Prevent double execution of commit epoch transaction by providing callback to ProcessBlock

* Simplify

goran-ethernal committed:
EVM-378 Use contracts api generated stubs in client code (#1166)

* Use contracts api for commitments

* Move ExitProof and StateSyncProof in another package

* Use state synced event from generated code

* Use generated structs for comit epoch (uptime)

* Rebase fix

* Use contracts api in checkpoints

* Use contracts api in bridge test

* Remove unnecessary logs

* Rebase fix

* Comments fix

* Comments fix

* Rebase fix

* Remove unused constants

* Comments fix

* Comments fix

* Rename functions and add comments

---------

Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>

yaziciahmet committed:
Fix EVM-372 & EVM-373 already existing directories and files are not checked for ownership and permissions (#1164)

* Implemented helper create file and dir functions

* Replace writeFile with CreateFileSafe

* replace mkdirall with CreateDirSafe

* change config file error message

* Add extra empty path guard

* Fix not overwriting files issue

* remove shouldNotExist parameter

* remove chmod update

* fix lint

* rename to save file

* comment fix

* comment fix

* one-liner return and handle errors

* update error message

Stefan-Ethernal committed:
Adapt go-ibft implementation for polybft (#1185)

* Fix hash calculation for proposal for go-ibft's update (#1065)

* Changing the signatures of the backend methods.

* Fixed all method signatures that expect a proposal.

* Verify committed seals created by ethereum block and round number

* Remove unnecessary fields before hash calculation for CS

* Fix missing field in packProposerSealIntoExtra

* Fix hash calculation for the block including round number

* Refactor hash calculation with round

* Fix build error test

* Revert "Refactor hash calculation with round"

This reverts commit b527ecb08e3d786ebf59d7988c86b934425b8582.

* Fix failed test

* Fix HasQuorum

* Rename method in IBFT Backend

* Fix method comment

* Rename IsValidBlock to IsValidProposal in IBFTBackend

* Fix parsing of RoundNumber in IBFTExtra

* Fix packCommittedSealsAndRoundNumberIntoExtra

* Revert packCommittedSealsAndRoundNumberIntoExtra

* Fix hash calculation for committed seals and revert interface of block verifier

* Update go-ibft

* Fix lint error

* Fix marshaling of IBFT Extra

* Fix minor issues

* Fix comment

---------

Co-authored-by: Lazar Travica <lazar.travica@mvpworkshop.co>

* Integrate updated go-ibft changes

* Error handling in BuildProposal

* Fix logs

* Add view instance to BuildProposal

* Update go-ibft version

* Formatting

---------

Co-authored-by: kourin <kourin.code@gmail.com>
Co-authored-by: Lazar Travica <lazar.travica@mvpworkshop.co>

ZeljkoBenovic committed:
[EVM-370] Added --insecure flag to both ibft and polybft (#1182)

* Added --insecure flag to both `ibft` and `polybft`
* Provide insecure flag in `polybft` e2e framework
* Provide insecure flag to `ibft` e2e framework
---------

Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>

goran-ethernal committed:
Reject contract creation transactions with large init data (#1181)

goran-ethernal committed:
Fix  validation of state transactions (#1186)

* Fix  validation of state transactions

* Add a comment

* Comments fix

* Add unit test

* Return error if two uptime transactions are in a block

---------

Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

goran-ethernal committed:
EVM-416 Use one struct for Exit and StateSync proofs in jsonrpc (#1170)

* Use one struct

* Fix state sync execution

* Remove abi functions from StateSyncProof

* Comments fix

* Comment
